Web: Add auto enrolling capabilities to RDS discover flow

kimlisa opened the pull request:
part of https://github.com/gravitational/teleport/issues/35434

requires 
- https://github.com/gravitational/teleport/pull/35930
- https://github.com/gravitational/teleport/pull/35899

recommend reviewing by commit

- cloud, RDS flow now defaults to auto-enrolling ~~(the toggler is pretty bad, it blends in with the background, i am working to update the design)~~ (after xmas breka)
- ~~self-hosted, we show a user a link to start a discovery service on their own for now~~ (in another PR)
- some tweaks
   - add more disabled state for RDS list
   - auto select integration if integration list has only one item
   - refactoring storybook/test

e2e testing in cloud staging:

https://www.loom.com/share/c68f402633134ca8be089ea01ac809d4?sid=fccb5441-c25c-4384-931e-203cdc166d2a

relevant stories:

http://127.0.0.1:9002/?path=/story/teleport-discover-database-enrollrds--instance-list-for-cloud
http://127.0.0.1:9002/?path=/story/teleport-discover-database-setupaccess--with-traits-aws-postgres-auto-enroll
http://127.0.0.1:9002/?path=/story/teleport-discover-database-deploy-auto--init-with-auto-enroll


changelog: Add auto-enrolling capabilities to RDS discover flow in the web UI

kimlisa committed:
Add discovery config endpoint

kimlisa committed:
Add autoDiscoveryConfig field to meta

kimlisa committed:
Hide custom label for auto discover

kimlisa committed:
Pull out re-usable parts for storybook fixture

kimlisa committed:
Only allow adding new traits when auto-enrolling

- Existing traits is not made editable
- Minor refactor: pull data from useDiscover context

kimlisa committed:
Auto select integration if only one result

kimlisa committed:
Skip test connection when auto enrolling

Refactor setup access story

kimlisa committed:
Add auto enroll toggler

kimlisa committed:
Create finish component for auto discover

kimlisa committed:
Add more disabled row states

kimlisa committed:
Add clarifying notes

gzdunek reviewed (commented):
First pass. I will try to test it tomorrow.

kimlisa commented on web/packages/teleport/src/Discover/Shared/SetupAccess/useUserTraits.ts:
```diff
@@ -222,14 +242,17 @@ export function useUserTraits(props: AgentStepProps) {
       default:
         throw new Error(
-          `useUserTraits.ts:updateResourceMetaDynamicTraits: resource kind ${props.resourceSpec.kind} is not handled`
+          `useUserTraits.ts:updateResourceMetaDynamicTraits: resource kind ${resourceSpec.kind} is not handled`
```
i can't use `assertUnreachable` because I didn't define all the cases for the enum i'm using (it'll complain until i define all the enum vals as cases). this function is more like "add more cases as we support more resources" and atm we only support setting traits for small select num resources.

kimlisa commented on web/packages/teleport/src/Discover/Database/SetupAccess/SetupAccess.tsx:
```diff
@@ -95,7 +94,17 @@ export function SetupAccess(props: State) {
+    // only supports one resource testing and auto enrolling
+    // enrolls resources > 1.
+    if (agentMeta.autoDiscoveryConfig) {
+      numStepsToSkip = 2;
```
so i messed up how this should work. originally nextStep() meant go to next step (which is currentStep +1). 
https://github.com/gravitational/teleport/blob/da589355d4ea55de276062db09f440c6fefdb2d6/web/packages/teleport/src/Discover/useDiscover.tsx#L321

then came the need to add ability to skip, and instead of passing in the number of steps to `skip`, i passed in the number of total increments. so `nextStep(2)` means skip only one step.

and this vairable name is misleaing, i renamed to `numStepsToIncrement`

kimlisa commented on web/packages/teleport/src/Discover/Database/CreateDatabase/useCreateDatabase.ts:
```diff
@@ -236,6 +243,63 @@ export function useCreateDatabase() {
+    let discoveryConfig = autoDiscoveryCfg;
+    try {
+      if (!discoveryConfig) {
+        discoveryConfig = await createDiscoveryConfig(clusterId, {
+          name: crypto.randomUUID(),
+          discoveryGroup: DISCOVERY_GROUP_CLOUD,
+          aws: [
+            {
+              types: ['rds'],
+              regions: [selectedRegion],
+              tags: { '*': ['*'] },
+              integration: (agentMeta as DbMeta).integration.name,
+            },
+          ],
+        });
+        setAutoDiscoveryCfg(discoveryConfig);
+      }
+    } catch (err) {
+      handleRequestError(err, 'failed to create discovery config: ');
+      return;
+    }
+
+    // See if we have existing database services that has asteriks or region
+    try {
+      const { services } = await ctx.databaseService.fetchDatabaseServices(
+        clusterId
+      );
```
hrm, should i try using `Promise.allSettled` instead? (though this may potentially take longer if there was an error with one)

it's because with `Promise.all` i don't think i can tell tell which promise failed and sometimes the backend returns confusing messages and we emit all errors:

```
function handleRequestError(err: Error, preErrMsg = '') {
    const message = getErrMessage(err);
    setAttempt({ status: 'failed', statusText: `${preErrMsg}${message}` });
    emitErrorEvent(`${preErrMsg}${message}`);
  }
```

kimlisa commented:
@gzdunek this PR isn't ready for testing, ran into some issues (or rather there are some things missing)

i'm debating keeping this PR as is, and create another PR based on this to complete, since this PR is already kinda noisy.

ryanclark commented on web/packages/teleport/src/Discover/Database/DeployService/AutoDeploy/AutoDeploy.tsx:
```diff
@@ -206,7 +214,15 @@ export function AutoDeploy({ toggleDeployMethod }: DeployServiceProp) {
 
             <StyledBox mb={5}>
               <Text bold>Step 4</Text>
```
Should this be step 3 if `wantAutoDiscover` is true?

ryanclark commented on web/packages/teleport/src/Discover/Database/EnrollRdsDatabase/AutoEnrollDialog.tsx:
```diff
@@ -0,0 +1,110 @@
+    // success
+    content = (
+      <>
+        <Flex mb={5} style={{ display: 'flex' }}>
```
Is `display: flex` needed here?

ryanclark commented on web/packages/teleport/src/Discover/Shared/Aws/RadioCell.tsx:
```diff
@@ -61,6 +61,7 @@ export function RadioCell<T>({
           onChange={() => onChange(item)}
           value={value}
           disabled={disabled}
+          data-testid="input-radio"
```
Can you use the `radio` role here instead of adding a testid?

ryanclark commented on web/packages/teleport/src/Discover/Shared/SetupAccess/useUserTraits.ts:
```diff
@@ -107,8 +115,11 @@ export function useUserTraits(props: AgentStepProps) {
-  function onProceed(traitOpts: Partial<Record<Trait, Option[]>>) {
-    switch (props.resourceSpec.kind) {
+  function onProceed(
+    traitOpts: Partial<Record<Trait, Option[]>>,
```
Is `Partial` needed here?

kimlisa replied:
yeah for each case we are partially (depending on resource) updating the fields defined here:
```ts
// UserTraits contain fields that define traits for local accounts.
export interface UserTraits {
  // logins is the list of logins that this user is allowed to
  // start SSH sessions with.
  logins: string[];
  // databaseUsers is the list of db usernames that this user is
  // allowed to open db connections as.
  databaseUsers: string[];
  // databaseNames is the list of db names that this user can connect to.
  databaseNames: string[];
  // kubeUsers is the list of allowed kube logins.
  kubeUsers: string[];
  // kubeGroups is the list of allowed kube groups for a kube cluster.
  kubeGroups: string[];
  // windowsLogins is the list of logins that this user
  // is allowed to start desktop sessions.
  windowsLogins: string[];
  // awsRoleArns is a list of aws roles this user is allowed to assume.
  awsRoleArns: string[];
}
```

gzdunek commented:
>(or rather there are some things missing)
>i'm debating keeping this PR as is, and create another PR based on this to complete, since this PR is already kinda noisy.

In that case, I think it makes sense to create another PR.

kimlisa committed:
Address CRs

kimlisa commented:
@ryanclark @gzdunek this is ready for another look. i decided to keep it all in one PR (it wasn't that much extra work, but maybe i'm just kidding myself...)

it's not ready to test yet since the backend portion is still WIP

and i'm so sorry that this PR became so noisy 😓

kimlisa commented:
drafted, another iteration requires changing some things again

kimlisa committed:
Finish auto discovery

- When enrolling and opting for auto discovery,
  we need to first collect all missing vpcs to determine
  if its required to deploy database agents, and then
  create a discovery config for a discovery service
- When auto deploying and opting for auto discovery,
  it is out scope for this iteration to allow user to
  define custom labels, select security groups, and
  "wait" for deploying all services to finish. Instead
  after deploy request is finished, we render a box
  that tells user it will take a minute to finish deploying
  and give a AWS dashboard link to view progress
- Setup access screen is the final step (skip test connection)
  and we render how to setup access for both mysql and postgres
  since auto discovery discovers all supported engine types.

kimlisa commented:
@gzdunek this is finally read for review. if you are interested in e2e testing, i have a cluster set up for it just ping me

you can also click around stories too, relavant stories:

http://127.0.0.1:9002/?path=/story/teleport-discover-database-enrollrds--instance-list-for-cloud
http://127.0.0.1:9002/?path=/story/teleport-discover-database-setupaccess--with-traits-aws-postgres-auto-enroll
http://127.0.0.1:9002/?path=/story/teleport-discover-database-deploy-auto--init-with-auto-enroll

e2e loom: https://www.loom.com/share/c68f402633134ca8be089ea01ac809d4?sid=fccb5441-c25c-4384-931e-203cdc166d2a

gzdunek commented on web/packages/teleport/src/Discover/Database/EnrollRdsDatabase/EnrollRdsDatabase.tsx:
```diff
@@ -222,14 +256,46 @@ export function EnrollRdsDatabase() {
+              disabled={tableData.items.length === 0}
+            >
+              <Box ml={2} mr={1}>
+                Auto-Enroll all Databases for selected region
```
`Auto-enroll`?

gzdunek commented on web/packages/teleport/src/Discover/Database/EnrollRdsDatabase/EnrollRdsDatabase.tsx:
```diff
@@ -240,24 +306,21 @@ export function EnrollRdsDatabase() {
+      {showTable && wantAutoDiscover && (
+        <Text mt={4} mb={-3}>
+          <b>Note:</b> Auto-Enroll will enroll <Mark>all</Mark> database engines
+          in this region (eg: postgres, mysql, aurora etc.)
```
```suggestion
          <b>Note:</b> Auto-enroll will enroll <Mark>all</Mark> database engines
          in this region (e.g. Postgres, MySQL, Aurora).
```

gzdunek reviewed (commented):
The code generally looks good.

>this is finally read for review. if you are interested in e2e testing, i have a cluster set up for it just ping me

Yeah, I'd like to test it!

kimlisa commented on web/packages/teleport/src/Discover/Database/EnrollRdsDatabase/EnrollRdsDatabase.tsx:
```diff
@@ -174,6 +192,87 @@ export function EnrollRdsDatabase() {
+        setRequiredVpcs(requiredVpcsAndSubnets);
+      } catch (err) {
+        handleAndEmitRequestError(err, {
+          preErrMsg: 'failed collecting vpc ids and its subnets: ',
```
i don't think our error messages from the back use capitalization?

kimlisa committed:
Address CR

gzdunek commented on web/packages/teleport/src/Discover/Database/EnrollRdsDatabase/EnrollRdsDatabase.tsx:
```diff
@@ -225,7 +225,7 @@ export function EnrollRdsDatabase() {
       } catch (err) {
         handleAndEmitRequestError(err, {
-          preErrMsg: 'failed collecting vpc ids and its subnets: ',
+          errorPrefix: 'failed collecting vpc ids and its subnets: ',
```
```suggestion
          errorPrefix: 'failed to collect vpc ids and its subnets: ',
```

gzdunek commented on web/packages/teleport/src/Discover/Database/EnrollRdsDatabase/EnrollRdsDatabase.tsx:
```diff
@@ -416,25 +415,25 @@ function getRdsEngineIdentifier(engine: DatabaseEngine): RdsEngineIdentifier {
       >
         <Box ml={2} mr={1}>
-          Auto-Enroll all Databases for selected region
+          Auto-enroll all Databases for selected region
```
I think `Databases` should be lowercase.

kimlisa commented:
ping @ryanclark

kimlisa committed:
Merge branch 'master' into lisa/rds-auto-discover
